core/src/main/java/org/apache/iceberg/avro/NameMappingWithAvroSchema.java

```diff
@@ -0,0 +1,119 @@
+
+    for (int i = 0; i < fieldResults.size(); i += 1) {
+      Types.NestedField field = iStruct.fields().get(i);
+      MappedFields result = fieldResults.get(i);
+      fields.add(MappedField.of(field.fieldId(), field.name(), result));
+    }
+
+    return MappedFields.of(fields);
+  }
+
+  @Override
+  public MappedFields union(Type iType, Schema union, List<MappedFields> optionResults) {
+    if (AvroSchemaUtil.isOptionSchema(union)) {
+      for (int i = 0; i < optionResults.size(); i += 1) {
+        if (union.getTypes().get(i).getType() != Schema.Type.NULL) {
+          return optionResults.get(i);
+        }
+      }
+    } else { // Complex union
+      Preconditions.checkArgument(
+          iType instanceof Types.StructType,
+          "Cannot visit invalid Iceberg type: %s for Avro complex union type: %s",
+          iType,
+          union);
+      Types.StructType iStruct = (Types.StructType) iType;
+      List<MappedField> fields = Lists.newArrayListWithExpectedSize(optionResults.size());
+      int index = 0;
+      for (Schema option : union.getTypes()) {
+        if (option.getType() != Schema.Type.NULL) {
+          // Check if current option is a named type, i.e., a RECORD, ENUM, or FIXED type. If so,
+          // use the full record name of the Avro type as the field name. Otherwise, use the Avro
+          // type toString().
+          if (option.getType() == Schema.Type.RECORD
+              || option.getType() == Schema.Type.ENUM
+              || option.getType() == Schema.Type.FIXED) {
+            fields.add(
+                MappedField.of(
+                    iStruct.fields().get(index).fieldId(),
+                    option.getFullName(),
+                    optionResults.get(index)));
+          } else {
+            fields.add(
+                MappedField.of(
+                    iStruct.fields().get(index).fieldId(),
+                    option.toString(),
+                    optionResults.get(index)));
+          }
+          // Both iStruct and optionResults do not contain an entry for the NULL type, so we need to
+          // increment i only
+          // when we encounter a non-NULL type.
+          index++;
+        }
+      }
+      return MappedFields.of(fields);
+    }
+    return null;
+  }
+
+  @Override
+  public MappedFields array(Types.ListType iList, Schema array, MappedFields elementResult) {
+    return MappedFields.of(MappedField.of(iList.elementId(), "element", elementResult));
+  }
+
+  @Override
+  public MappedFields map(
+      Types.MapType iMap, Schema map, MappedFields keyResult, MappedFields valueResult) {
+    return MappedFields.of(
+        MappedField.of(iMap.keyId(), "key", keyResult),
+        MappedField.of(iMap.valueId(), "value", valueResult));
+  }
+
+  @Override
+  public MappedFields map(Types.MapType iMap, Schema map, MappedFields valueResult) {
+    return MappedFields.of(MappedField.of(iMap.valueId(), "value", valueResult));
```

Comment:
This still needs to create a mapping for the key field, right? The Iceberg map will have a key field with a String type. There won't be a nested mapping because the key is a String, but there should be a mapping from `iMap.keyId()` to `"key"`

Reply:
Added map key mapping.